clustered/server/src/main/java/org/ehcache/clustered/server/offheap/DefaultOffHeapChainHeaderExtension.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.server.offheap;
+
+import org.terracotta.offheapstore.paging.OffHeapStorageArea;
+import org.terracotta.offheapstore.storage.StorageEngine;
+import org.terracotta.offheapstore.storage.portability.WriteContext;
+
+import java.nio.ByteBuffer;
+
+public class DefaultOffHeapChainHeaderExtension implements OffHeapChainExtensionHeader {
+  private final OffHeapStorageArea storage;
+  private final int extendedChainHeaderSize;
+  private final int chainHeaderSize;
+
+  private StorageEngine.Owner owner;
+
+  DefaultOffHeapChainHeaderExtension(OffHeapStorageArea storage, int extendedChainHeaderSize, int chainHeaderSize) {
+    this.storage = storage;
+    this.extendedChainHeaderSize = extendedChainHeaderSize;
+    this.chainHeaderSize = chainHeaderSize;
+  }
+
+  @Override
+  public ByteBuffer readXtnHeader(long xtnAddress) {
+    return storage.readBuffer(xtnAddress, extendedChainHeaderSize);
+  }
+
+  @Override
+  public WriteContext getXtnWriteContext(final long xtnAddress) {
+    return new WriteContext() {
+
+      @Override
+      public void setLong(int offset, long value) {
+        if (offset < 0 || offset > extendedChainHeaderSize) {
+          throw new IllegalArgumentException();
+        } else {
+          storage.writeLong(xtnAddress + offset, value);
+        }
+      }
+
+      @Override
+      public void flush() {
+        //no-op
+      }
+    };
+  }
+
+  public Long extractXtnAddress(int hash, ByteBuffer keyBuffer) {
+    Long encoding = owner.getEncodingForHashAndBinary(hash, keyBuffer);
```

Comment:
Actually bind is guaranteed to happen before a storage engine is used by the owning map. there is a lot of places where it will fail with null pointer exception if it does not.